modules/importer/src/server/mod.rs

```diff
@@ -56,84 +61,52 @@ struct Server {
-
-                service.update_start(&importer.name, None).await?;
-
-                // record timestamp before processing, so that we can use it as "since" marker
-                let last_run = OffsetDateTime::now_utc();
-
-                log::info!("Starting run: {}", importer.name);
 
-                let context = ServiceRunContext::new(service.clone(), importer.name.clone());
-
-                let runner = ImportRunner {
-                    db: self.db.clone(),
-                    storage: self.storage.clone(),
-                    working_dir: self.working_dir.clone(),
-                    analysis: self.analysis.clone(),
-                };
-
-                let (last_error, report, continuation) = match runner
-                    .run_once(
-                        context,
-                        importer.data.configuration,
-                        importer.data.last_success,
-                        importer.data.continuation,
+            // Remove jobs that are finished
+            runs.retain(|_, job| !job.is_finished());
+
+            // Asynchronously fire off new jobs subject to max concurrency
+            let todo: Vec<_> = service
+                .list()
+                .await?
+                .into_iter()
+                .filter(|i| {
+                    !(runs.contains_key(&i.name) || i.data.configuration.disabled || can_wait(i))
+                })
+                .take(self.concurrency - runs.len())
+                .map(|i| {
+                    (
+                        i.name.clone(),
+                        spawn_local(import(runner.clone(), i, service.clone())),
                     )
-                    .await
-                {
-                    Ok(RunOutput {
-                        report,
-                        continuation,
-                    }) => (None, Some(report), continuation),
-                    Err(ScannerError::Normal {
-                        err,
-                        output:
-                            RunOutput {
-                                report,
-                                continuation,
-                            },
-                    }) => (Some(err.to_string()), Some(report), continuation),
-                    Err(ScannerError::Critical(err)) => (Some(err.to_string()), None, None),
-                };
-
-                log::info!("Import run complete: {last_error:?}");
+                })
+                .collect();
 
-                service
-                    .update_finish(
-                        &importer.name,
-                        None,
-                        last_run,
-                        last_error,
-                        continuation,
-                        report.and_then(|report| serde_json::to_value(report).ok()),
-                    )
-                    .await?;
-            }
+            // Add them to our list of currently running jobs
+            runs.extend(todo.into_iter());
```

Comment:
could the `todo` list contain an importer already in the `runs` list?

Reply:
The check on line 97, `runs.contains_key(&i.name)` is meant to prevent that.